pkg/telemetry/platform.go

```diff
@@ -0,0 +1,99 @@
+package telemetry
+
+import (
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+)
+
+const (
+	openshiftIdentifier = "node.openshift.io/os_id"
+	k3sIdentifier       = "k3s"
+	awsIdentifier       = "aws"
+	gkeIdentifier       = "gce"
+	azureIdentifier     = "azure"
+	kindIdentifier      = "kind"
+	rancherIdentifier   = "cattle-system"
+)
+
+func CollectK8sPlatform(node v1.Node, namespaces v1.NamespaceList) string {
+	if result := isMultiplePlatforms(node, namespaces); result != "" {
+		return result
+	}
+
+	if isAWSPlatform(node) {
```

Comment:
the remaining functions can be simplified with something like below:
```go
const (
	gkeID = "gke"
	eksID = "eks"
)

var mapProviderIDToPlatform = map[string]ClusterPlatform{
	gkeID: ClusterPlatformGKE,
	eksID: ClusterPlatformEKS,
        // ...
}

func getPlatformFromProviderID(providerID string) ClusterPlatform {
	for id, platform := range mapProviderIDToPlatform {
		if strings.HasPrefix(providerID, id) {
			return platform
		}
	}
	return ""
}
```

Reply:
I ran out of time yesterday to add this, but going off of this [github](https://github.com/cisco-open/cluster-registry-controller/blob/3086c2cb8cd559891bc19d746b2f1f89721a77ed/pkg/clustermeta/distributions.go#L120) which does something similar, you can see at the line I linked, for some of the providers they check both the provderID if applicable AND a `node.Label` as you can see for gke.

They also instead check for node.Annotations for k3s and rancher.


I chose deliberately to only check for providerID on platforms that supply it as I think it
- Simplifies the code
- Is the most "correct" way to check for the provider as that is a field specifically meant for the cloud provider to put identification in
- Is least likely to get changed

I want to pose a question, do you think just checking the providerID is satisfactory? I think it is because especially for telemetry data, I don't think we need to be perfect and if for some reason a platform gets mislabeled it will most likely be an edge case which shouldn't affect telemetry data results as we will have a room for error when analyzing the data. 

If just checking the providerID is fine then I'll move forward with this, else I left each platform as a separate function since we could do unique checks for each one if necessary.

Reply:
There is also this [github](https://github.com/litmuschaos/litmusctl/blob/95fad15b010a9a53afec315a04f7b0c4a636f675/pkg/infra_ops/platform.go#L123)  which only does check providerID check.

Reply:
checking providerID only is reasonable, unless there is some strong evidence it's not gonna work

Reply:
@shaun-nx @jjngx Do you two have any thoughts on this?

Reply:
@bjee19 from reading your points I also think the providerID is sufficient for now. I also agree that it does simply the implementation. 

I do have one concern.
Right now, if the `providerID` does not match, we return an empty string to indicate this.
I feel like it would make sense to just return the original `providerID`

For example, if we have a providerID of `my-new-provider://local-node`, this wont match. So we can choose to return this full string and send that as the deployment platform.

The reason I think we should do this is because if a ProviderID does ever change, we will see what that new value is in our telemetry output. If we see just an empty string, we don't know what the value of the ID was, so we don't know what we missed.

Curious to know what you folks think @bjee19 @pleshakov @jjngx

Reply:
@shaun-nx I do like the thought of this, our story decided on marking it specifically as `other`, so I'll ping product. @mpstefan. 

My one issue with this is: if we return the original `providerID` because it didn't match any of the ones we have listed, I would personally count it as an edge case. This could cause an issue with reporting PII information because there is no guarantee that they didn't set the `providerID` output to something like `john-smith-address-here://local-node` and we would have no checks on it. 

The second concern on if the `providerID` does change for any of the platforms we won't see what the new value is on our telemetry output is valid. And my initial thoughts are potentially creating a test in the future that monitors each cluster and ensures the providerID is still what we think it is, however that is a lot and might be very much unnecessary. I can't really see a way around this.

Reply:
I don't think we need to think TOO hard about this. If we can get an accurate string that represents the platform for 90% of cases just by grabbing the providerID, I say we grab that. I'd be interested in breaking up the "other" label to see if there are any trends that we weren't aware of - I know OpenShift is going to be one of these, seeing as how it doesn't have a providerID. We may need to do something about that in future...

For PII, someone would need to create their own version of Kubernetes and put their name or address or some other PII in the providerID field. That is obviously very far from what the field was intended for and I think the risk is extremely low. If, on the off chance, we do end up picking up something troubling, we can purge it before we store it on the data side then prioritize a fix in NGF to never have it sent.

Reply:
Openshift won't be in the "other" since we're able to check that, same with Rancher. 

But I'll do what Shaun suggested and break up the "other" category by reporting the original `providerID`!

Reply:
Consider the following provider IDs:
```
aks:
providerID: azure:///subscriptions/…

gke:
providerID: gce:///us-central1-c/
````

In Azure, it gives the subscription id, right?

Is it the name of the project on GCE?

Those are private details imho.

So if there is some provider that we don't support, they might include some private details there - thus we should not report it.

Reply:
I agree with you and I was planning on only sending the `ProviderName` section of the providerID

```
// ID of the node assigned by the cloud provider in the format: <ProviderName>://<ProviderSpecificNodeID>
// +optional
ProviderID string `json:"providerID,omitempty" protobuf:"bytes,3,opt,name=providerID"`
```

Which I think aligns more with your view @pleshakov